File: api/src/main/java/com/datastrato/gravitino/file/BaseFilesetDataOperationCtx.java
/*
 * Copyright 2024 Datastrato Pvt Ltd.
 * This software is licensed under the Apache License version 2.
 */
package com.datastrato.gravitino.file;

import com.google.common.base.Objects;
import com.google.common.base.Preconditions;
import java.util.Map;

/** Base implementation of {@link FilesetDataOperationCtx}. */
public class BaseFilesetDataOperationCtx implements FilesetDataOperationCtx {
  private String subPath;
  private FilesetDataOperation operation;
  private ClientType clientType;
  private String ip;
  private SourceEngineType sourceEngineType;
  private String appId;
  private Map<String, String> extraInfo;

  @Override
  public String subPath() {
    return subPath;
  }

  @Override
  public FilesetDataOperation operation() {
    return operation;
  }

  @Override
  public ClientType clientType() {
    return clientType;
  }

  @Override
  public String ip() {
    return ip;
  }

  @Override
  public SourceEngineType sourceEngineType() {
    return sourceEngineType;
  }

  @Override
  public String appId() {
    return appId;
  }

  @Override
  public Map<String, String> extraInfo() {
    return extraInfo;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (!(o instanceof BaseFilesetDataOperationCtx)) return false;
    BaseFilesetDataOperationCtx that = (BaseFilesetDataOperationCtx) o;
    return Objects.equal(subPath, that.subPath)
        && operation == that.operation
        && clientType == that.clientType
        && Objects.equal(ip, that.ip)
        && sourceEngineType == that.sourceEngineType
        && Objects.equal(appId, that.appId)
        && Objects.equal(extraInfo, that.extraInfo);
  }

  @Override
  public int hashCode() {
    return Objects.hashCode(subPath, operation, clientType, ip, sourceEngineType, appId, extraInfo);
  }

  public static class Builder {
    private BaseFilesetDataOperationCtx context;
    /** Creates a new instance of {@link BaseFilesetDataOperationCtx.Builder}. */
    private Builder() {
      context = new BaseFilesetDataOperationCtx();
    }

    public BaseFilesetDataOperationCtx.Builder withSubPath(String subPath) {
      context.subPath = subPath;
      return this;
    }

    public BaseFilesetDataOperationCtx.Builder withOperation(FilesetDataOperation operation) {
      context.operation = operation;
      return this;
    }

    public BaseFilesetDataOperationCtx.Builder withClientType(ClientType clientType) {
      context.clientType = clientType;
      return this;
    }

    public BaseFilesetDataOperationCtx.Builder withIp(String ip) {
      context.ip = ip;
      return this;
    }

    public BaseFilesetDataOperationCtx.Builder withSourceEngineType(
        SourceEngineType sourceEngineType) {
      context.sourceEngineType = sourceEngineType;
      return this;
    }

    public BaseFilesetDataOperationCtx.Builder withAppId(String appId) {
      context.appId = appId;
      return this;
    }

    public BaseFilesetDataOperationCtx.Builder withExtraInfo(Map<String, String> extraInfo) {
      context.extraInfo = extraInfo;
      return this;
    }

    private void validate() {
      Preconditions.checkArgument(context.subPath != null, "subPath cannot be null");
      Preconditions.checkArgument(context.operation != null, "operation is required");
      Preconditions.checkArgument(context.clientType != null, "clientType is required");
      Preconditions.checkArgument(context.sourceEngineType != null, "sourceEngineType is required");
    }

    public BaseFilesetDataOperationCtx build() {
      validate();
      return context;
    }
  }

  /**
   * Creates a new instance of {@link Builder}.
   *
   * @return The new instance.
   */
  public static Builder builder() {
    return new Builder();
  }
}

File: clients/client-python/gravitino/api/base_fileset_data_operation_ctx.py
"""
Copyright 2024 Datastrato Pvt Ltd.
This software is licensed under the Apache License version 2.
"""

from typing import Dict

from gravitino.api.client_type import ClientType
from gravitino.api.fileset_data_operation import FilesetDataOperation
from gravitino.api.fileset_data_operation_ctx import FilesetDataOperationCtx
from gravitino.api.source_engine_type import SourceEngineType


class BaseFilesetDataOperationCtx(FilesetDataOperationCtx):
    """Base implementation of FilesetDataOperationCtx."""

    _sub_path: str
    _operation: FilesetDataOperation
    _client_type: ClientType
    _ip: str
    _source_engine_type: SourceEngineType
    _app_id: str
    _extra_info: Dict[str, str]

    def __init__(
        self,
        sub_path: str,
        operation: FilesetDataOperation,
        client_type: ClientType,
        ip: str,
        source_engine_type: SourceEngineType,
        app_id: str,
        extra_info: Dict[str, str] = None,
    ):
        assert sub_path is not None
        assert operation is not None
        assert client_type is not None
        assert source_engine_type is not None
        self._sub_path = sub_path
        self._operation = operation
        self._client_type = client_type
        self._ip = ip
        self._source_engine_type = source_engine_type
        self._app_id = app_id
        self._extra_info = extra_info

    def sub_path(self) -> str:
        return self._sub_path

    def operation(self) -> FilesetDataOperation:
        return self._operation

    def client_type(self) -> ClientType:
        return self._client_type

    def ip(self) -> str:
        return self._ip

    def source_engine_type(self) -> SourceEngineType:
        return self._source_engine_type

    def app_id(self) -> str:
        return self._app_id

    def extra_info(self) -> Dict[str, str]:
        return self._extra_info
